### PR TITLE
[SYCLomatic] Support the migration of 5 cuda API about L2 Cache

### PR DESCRIPTION
**Pull request by wenhuiNi**
Support the migration of 5 cuda API about L2cahe by removing the call.
APIName: `cudaCtxResetPersistingL2Cache, cuCtxResetPersistingL2Cache, cudaStreamSetAttribute, cudaStreamGetAttribute, cudaFuncSetAttribute`
Signed-off-by: Ni, Wenhui <wenhui.ni@intel.com>

### CODE REVIEW

**Review by zhiweij1 — APPROVED** [review]

File: clang/lib/DPCT/APINames_removed.inc [hunk 1]
```diff
@@ -13,6 +13,12 @@ ENTRY(cudaDeviceSetLimit,              "SYCL currently does not support setting
 ENTRY(cudaThreadSetLimit,              "SYCL currently does not support setting resource limits on devices.")
 ENTRY(cudaDeviceSetCacheConfig,        "SYCL currently does not support setting cache config on devices.")
 ENTRY(cudaDeviceGetCacheConfig,        "SYCL currently does not support setting cache config on devices.")
+ENTRY(cudaCtxResetPersistingL2Cache,   "SYCL currently does not support setting cache config on devices.")
+ENTRY(cuCtxResetPersistingL2Cache,     "SYCL currently does not support setting cache config on devices.")
+ENTRY(cudaStreamSetAttribute,          "SYCL currently does not support setting cache config on devices.")
+ENTRY(cudaStreamGetAttribute,          "SYCL currently does not support setting cache config on devices.")
+ENTRY(cudaFuncSetAttribute,          "SYCL currently does not support setting cache config on devices.")
```
Comment (zhiweij1) [inline clang/lib/DPCT/APINames_removed.inc]:
```suggestion
ENTRY(cudaFuncSetAttribute,            "SYCL currently does not support setting cache config on devices.")
```

**Review by zhiweij1 — CHANGES_REQUESTED** [review]
Some format issues.

**Review by zhiweij1 — APPROVED** [review]

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 2]
```diff
@@ -14657,3 +14663,28 @@ void CudaUuidRule::runRule(
...
+  MF.addMatcher(binaryOperator(allOf(isAssignmentOperator(), hasLHS(
+      hasDescendant(declRefExpr(to(varDecl(hasType(qualType(hasDeclaration(
+                      typedefDecl(hasAnyName("cudaLaunchAttributeValue"))))))))))))
+          .bind("AssignStmtRemove"),
+      this);
+}
+
+void TypeRemoveRule::runRule(
+    const ast_matchers::MatchFinder::MatchResult &Result) {
+  if (auto VD = getNodeAsType<VarDecl>(Result, "VarDeclRemove")) {
```
Comment (yihwang-nv) [inline clang/lib/DPCT/ASTTraversal.cpp]: Typically we don't do curleys on a single-statement.
```
[Don’t Use Braces on Simple Single-Statement Bodies of if/else/loop Statements](https://llvm.org/docs/CodingStandards.html#id59)
```

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 3]
```diff
@@ -14657,3 +14663,28 @@ void CudaUuidRule::runRule(
...
+    emplaceTransformation(ReplaceVarDecl::getVarDeclReplacement(VD, ""));
+  }
+  if (auto BO = getNodeAsType<BinaryOperator>(Result, "AssignStmtRemove")) {
```
Comment (yihwang-nv) [inline clang/lib/DPCT/ASTTraversal.cpp]: likewise

**Review by yihwang-nv — APPROVED** [review]

File: clang/test/dpct/cache_remove.cu [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      transformed_tensor_size * 2 + pre_transform_tensor_size;
+  float *aptr;
+  // CHECK: DPCT1007:{{[0-9]+}}: Migration of cudaLaunchAttributeValue is not supported.
+  cudaStreamAttrValue stream_attribute = {};
+  stream_attribute.accessPolicyWindow.base_ptr = aptr;
+  stream_attribute.accessPolicyWindow.num_bytes = res_block_mem;
+  stream_attribute.accessPolicyWindow.hitRatio = 1.0f;
+  stream_attribute.accessPolicyWindow.hitProp = cudaAccessPropertyPersisting;
+  stream_attribute.accessPolicyWindow.missProp = cudaAccessPropertyStreaming;
+  // CHECK: DPCT1007:{{[0-9]+}}: Migration of cudaStreamSetAttribute is not supported.
```
Comment (zhimingwang36) [inline clang/test/dpct/cache_remove.cu]: can remove all the cudaStreamSetAttribute(..) here?

File: clang/test/dpct/cache_remove.cu [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      stream, cudaStreamAttributeAccessPolicyWindow, &stream_attribute);
+  // CHECK: DPCT1026:{{[0-9]+}}: The call to cudaCtxResetPersistingL2Cache was removed because SYCL currently does not support setting cache config on devices.
+  cudaCtxResetPersistingL2Cache();
+  // CHECK: DPCT1026:{{[0-9]+}}: The call to cuCtxResetPersistingL2Cache was removed because SYCL currently does not support setting cache config on devices.
+  cuCtxResetPersistingL2Cache();
+  // CHECK: DPCT1026:{{[0-9]+}}: The call to cudaFuncSetAttribute was removed because SYCL currently does not support configuring shared emory on devices.
+  cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, 8 * 8 * 8);
+  // CHECK: DPCT1026:{{[0-9]+}}: The call to cudaFuncSetAttribute was removed because SYCL currently does not support configuring shared emory on devices.
+  cudaFuncSetAttribute(kernel, cudaFuncAttributePreferredSharedMemoryCarveout, 8 * 8 * 8);
+  // CHECK: DPCT1007:{{[0-9]+}}: Migration of cudaFuncSetAttribute is not supported.
```
Comment (zhimingwang36) [inline clang/test/dpct/cache_remove.cu]: why some 1007 and other 1026?

Comment (wenhuiNi) [reply clang/test/dpct/cache_remove.cu]: > why some 1007 and other 1026?
It depends on the usage of the function call. If function call be used for setting cache or shared memory config, we will report 1026 or 1027. In other case, we consider it as an unprocessed CUDA API call, report 1007 warning.

File: clang/test/dpct/cache_remove.cu (hunk 4 above)
Comment (wenhuiNi) [reply clang/test/dpct/cache_remove.cu]: > can remove all the cudaStreamSetAttribute(..) here?

no, cudaStreamSetAttribute be used for setting stream attribute include cudaLaunchAttributeAccessPolicyWindow for cache, Synchronization Policy, Priority.
So, we need check the 2nd param value of cudaStreamSetAttribute to determine whether the call is used to setting cache config.

Only this call be used to setting cudaStream cache config, we will remove this call.

File: clang/test/dpct/cache_remove.cu (hunk 5 above)
Comment (wenhuiNi) [reply clang/test/dpct/cache_remove.cu]: Fixed, all of them will be remove and report 1026

**Review by ziranzha — APPROVED** [review]

File: clang/lib/DPCT/APINames.inc [hunk 6]
```diff
@@ -95,14 +95,14 @@ ENTRY(cudaStreamCreateWithFlags, cudaStreamCreateWithFlags, true, NO_FLAG, P0, "
 ENTRY(cudaStreamCreateWithPriority, cudaStreamCreateWithPriority, true, NO_FLAG, P0, "Successful: DPCT1025")
 ENTRY(cudaStreamDestroy, cudaStreamDestroy, true, NO_FLAG, P0, "Successful")
 ENTRY(cudaStreamEndCapture, cudaStreamEndCapture, true, API_CALL_REMOVED, P4, "DPCT1026/DPCT1027")
-ENTRY(cudaStreamGetAttribute, cudaStreamGetAttribute, false, NO_FLAG, P7, "comment")
+ENTRY(cudaStreamGetAttribute, cudaStreamGetAttribute, false, NO_FLAG, P7, "DPCT1026/DPCT1027")
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APINames.inc]: false ==> true?

File: docs/dev_guide/diagnostic_ref/dpct1026.rst [hunk 7]
```diff
@@ -46,6 +46,10 @@ For example, this original CUDA\* code:
    __device__ void bar(int *a) {
      __ldg(a);
    }
```
Comment (zhimingwang36) [inline docs/dev_guide/diagnostic_ref/dpct1026.rst]: Please split the DOC change files into separate PR.

File: clang/test/dpct/cache_remove.cu [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+int main()
+{
+  cudaStream_t stream;
+  const int pre_transform_tensor_size =
+      8 * 8 * 8;
+  const int transformed_tensor_size = pre_transform_tensor_size * 36 / 16;
+  const int res_block_mem =
+      transformed_tensor_size * 2 + pre_transform_tensor_size;
+  float *aptr;
+  // CHECK: DPCT1007:{{[0-9]+}}: Migration of cudaLaunchAttributeValue is not supported.
```
Comment (zhimingwang36) [inline clang/test/dpct/cache_remove.cu]: #define cudaStreamAttrValue cudaLaunchAttributeValue
In L21,  it's better to use cudaStreamAttrValue instead or else user will be mislead.

File: clang/test/dpct/cache_remove.cu [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  // CHECK: DPCT1026:{{[0-9]+}}: The call to cudaStreamGetAttribute was removed because SYCL currently does not support setting cache config on devices.
+  cudaStreamGetAttribute(
+      stream, cudaStreamAttributeAccessPolicyWindow, &stream_attribute);
+  // CHECK: DPCT1026:{{[0-9]+}}: The call to cudaCtxResetPersistingL2Cache was removed because SYCL currently does not support setting cache config on devices.
+  cudaCtxResetPersistingL2Cache();
+  // CHECK: DPCT1026:{{[0-9]+}}: The call to cuCtxResetPersistingL2Cache was removed because SYCL currently does not support setting cache config on devices.
+  cuCtxResetPersistingL2Cache();
+  // CHECK: DPCT1026:{{[0-9]+}}: The call to cudaFuncSetAttribute was removed because SYCL currently does not support configuring shared emory on devices.
+  cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, 8 * 8 * 8);
+  // CHECK: DPCT1026:{{[0-9]+}}: The call to cudaFuncSetAttribute was removed because SYCL currently does not support configuring shared emory on devices.
```
Comment (zhimingwang36) [inline clang/test/dpct/cache_remove.cu]:
```suggestion
  // CHECK: DPCT1026:{{[0-9]+}}: The call to cudaFuncSetAttribute was removed because SYCL currently does not corresponding setting.
```

File: clang/test/dpct/cache_remove.cu [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  cudaFuncSetAttribute(kernel, cudaFuncAttributePreferredSharedMemoryCarveout, 8 * 8 * 8);
+  // CHECK: DPCT1026:{{[0-9]+}}: The call to cudaFuncSetAttribute was removed because SYCL currently does not support adjusting attribute for kernel function.
```
Comment (zhimingwang36) [inline clang/test/dpct/cache_remove.cu]:
```suggestion
  // CHECK: DPCT1026:{{[0-9]+}}: The call to cudaFuncSetAttribute was removed because SYCL currently does not support corresponding setting.
```

File: clang/test/dpct/cache_remove.cu (hunk 8 above)
Comment (wenhuiNi) [reply clang/test/dpct/cache_remove.cu]: fixed. now, we will use cudaStreamAttrValue in warning msg

**Review by ziranzha — APPROVED** [review]
